### PR TITLE
refactor: Ensure Celery leverages the Flask-SQLAlchemy session

### PR DESCRIPTION
**Pull request by john-bodley**
### SUMMARY

As part of SIP-99 (specifically [SIP-99A](https://github.com/apache/superset/issues/25107) and [SIP-99B](https://github.com/apache/superset/issues/25108)) in order to ensure a consistent "unit of work"—via a single atomic unit—all operations should be associated with the same Flask-SQLAlchemy session.

The Flask-SQLAlchemy extension provides a scoped session (on a per request basis) with the necessary oversight, i.e., the session is closed after the request is complete which aids with connection pool management. 

Historically Celery tasks have defined their own scoped session (with the option to use connection pooling) which needed to managed independently which added unnecessary code bloat and complexity and likely violated the "unit of work" construct if operations were leveraging both the Flask-SQLAlchemy and Celery sessions. Per [this](https://gist.github.com/twolfson/a1b329e9353f9b575131) post it seems like Celery can piggyback off of the Flask-SQLAlchemy session so long as setup/teardown is handled correctly.  There is actually an example of this in the official Flask documentation per the [Celery with Flask](https://flask.palletsprojects.com/en/2.3.x/patterns/celery/) document where it references the use of the Flask-SQLAlchemy session (`db.session`).

This PR removes the need for a Celery specific session which is a step towards the goal of having all database operations (outside of the Alembic migrations) handled by the global `db.session` which is a necessary requirement in order for us to achieve the goal of an atomic unit of work.

The one wrinkle with this approach is https://github.com/apache/superset/pull/10819 which explicitly leveraged a `NullPool` (as opposed to `QueuePool`—if defined via the `SQLALCHEMY_ENGINE_OPTIONS ` configuration) to avoid using same session across multiple celery workers, however since we're tearing down the session a per task basis where, per [here](https://stackoverflow.com/questions/12044776/how-to-use-flask-sqlalchemy-in-a-celery-task/14146403#14146403), it states,

> Flask SQLAlchemy will automatically create new sessions for you from scoped session factory, given that we are maintaining the same app context, this ensures tasks have a fresh session (e.g. session errors won't propagate across tasks).

i.e., we should no longer be experiencing the connection bleeding issue and thus can leverage the efficiency of connection pooling which standardizing the code.

### TESTING INSTRUCTIONS

CI. Additionally I wasn't able to repo the issue mentioned in https://github.com/apache/superset/issues/10530.

### ADDITIONAL INFORMATION

- [x] Has associated issue: https://github.com/apache/superset/issues/25107
- [ ] Required feature flags:
- [ ] Changes UI
- [ ] Includes DB Migration (follow approval process in [SIP-59](https://github.com/apache/superset/issues/13351))
  - [ ] Migration is atomic, supports rollback & is backwards-compatible
  - [ ] Confirm DB migration upgrade and downgrade tested
  - [ ] Runtime estimates and downtime expectations provided
- [ ] Introduces new feature or API
- [ ] Removes existing feature or API

### CODE REVIEW

File: superset/sql_lab.py [hunk 1]
```diff
@@ -176,30 +170,27 @@ def get_sql_results(  # pylint: disable=too-many-arguments
...
-                    start_time=start_time,
-                    expand_data=expand_data,
-                    log_params=log_params,
-                )
-            except Exception as ex:  # pylint: disable=broad-except
-                logger.debug("Query %d: %s", query_id, ex)
-                stats_logger.incr("error_sqllab_unhandled")
-                query = get_query(query_id, session)
-                return handle_query_error(ex, query, session)
+    with override_user(security_manager.find_user(username)):
```
Comment (john-bodley) [inline superset/sql_lab.py:173]: Pretty much the same code as previously without the outer `with` block.

File: superset/tasks/cache.py [hunk 2]
```diff
@@ -130,28 +123,24 @@ def __init__(self, top_n: int = 5, since: str = "7 days ago") -> None:
         self.since = parse_human_datetime(since) if since else None
 
     def get_payloads(self) -> list[dict[str, int]]:
-        payloads = []
-        session = db.create_scoped_session()
+        records = (
```
Comment (john-bodley) [inline superset/tasks/cache.py:126]: Pretty much the same code as previously without the `try` block.

File: superset/tasks/cache.py [hunk 3]
```diff
@@ -178,48 +167,44 @@ def __init__(self, tags: Optional[list[str]] = None) -> None:
...
-
-            # add dashboards that are tagged
-            tagged_objects = (
-                session.query(TaggedObject)
-                .filter(
-                    and_(
-                        TaggedObject.object_type == "dashboard",
-                        TaggedObject.tag_id.in_(tag_ids),
-                    )
+        tags = db.session.query(Tag).filter(Tag.name.in_(self.tags)).all()
```
Comment (john-bodley) [inline superset/tasks/cache.py:170]: Pretty much the same code as previously without the `try` block.

File: superset/tasks/scheduler.py [hunk 4]
```diff
@@ -46,35 +45,32 @@ def scheduler() -> None:
...
-        triggered_at = (
-            datetime.fromisoformat(scheduler.request.expires)
-            - app.config["CELERY_BEAT_SCHEDULER_EXPIRES"]
-            if scheduler.request.expires
-            else datetime.utcnow()
-        )
-        for active_schedule in active_schedules:
-            for schedule in cron_schedule_window(
-                triggered_at, active_schedule.crontab, active_schedule.timezone
+    active_schedules = ReportScheduleDAO.find_active()
```
Comment (john-bodley) [inline superset/tasks/scheduler.py:48]: Pretty much the same code as previously without the outer `with` block.

**Comment by john-bodley** [comment]
ping @michael-s-molina @villebro

**Comment by michael-s-molina** [comment]
@john-bodley I think it would be great to label this PR with v4.0 and merge it during the breaking window to reuse the test/stabilization efforts that will occur during that period.

**Comment by john-bodley** [comment]
Per,

> @john-bodley I think it would be great to label this PR with v4.0 and merge it during the breaking window to reuse the test/stabilization efforts that will occur during that period.

as discussed with @michael-s-molina, though this is technically a non-breaking change, it seems prudent (from a safety perspective) to hold off merging this until the v4.0 breaking window.

File: superset/tasks/celery_app.py [hunk 5]
```diff
@@ -43,3 +43,27 @@ def reset_db_connection_pool(**kwargs: Any) -> None:  # pylint: disable=unused-a
...
+@task_postrun.connect
+def teardown(  # pylint: disable=unused-argument
+    retval: Any,
+    *args: Any,
+    **kwargs: Any,
+) -> None:
+    """
+    After each Celery task teardown the Flask-SQLAlchemy session.
+
+    Note for non eagar requests Flask-SQLAlchemy will perform the teardown.
```
Comment (michael-s-molina) [inline superset/tasks/celery_app.py:57]:
```suggestion
    Note for non eager requests Flask-SQLAlchemy will perform the teardown.
```

**Review by michael-s-molina — APPROVED** [review]
LGTM. I'm assuming that removing unnecessary `commit` operations is not in the scope of this PR.

**Pull request merged**